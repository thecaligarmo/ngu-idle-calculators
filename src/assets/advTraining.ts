import { Stat } from "./stat"
import Resource, { ResourceContainer } from "./resource"

export class AdvTraining extends Resource{
    updateStats() {
        for (var prop of Object.keys(this.base)) {
            if (this.key == 'adventureToughness' || this.key == 'adventurePower') {
                this[prop] = this.level ** 0.4 * 10
            } else if (this.key =='blockDamage') {
                this[prop] = Math.floor( (this.level + 50) / (this.level + 100))
            } else {
                this[prop] = this.level * this.base[prop]
            }
        }
    }
}


export const ADVTRAININGLIST = [
    new AdvTraining(0, 'adventureToughness', 'The Fu Manchu', 0, [[Stat.TOUGHNESS, 1]]),
    new AdvTraining(1, 'adventurePower', 'The Neckbeard', 0, [[Stat.POWER, 1]]),
    new AdvTraining(2, 'blockDamage', 'The Revese Hitler', 0, []),
    new AdvTraining(3, 'wandoosEnergyDump', 'The Beard Cage', 0, [[Stat.WANDOOS_SPEED, 1]]),
    new AdvTraining(4, 'wandoosMagicDump', 'The LadyBeard', 0, [[Stat.WANDOOS_SPEED, 1]]),
]

export var ADVTRAININGS = new ResourceContainer(ADVTRAININGLIST);