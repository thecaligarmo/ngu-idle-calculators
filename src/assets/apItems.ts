import { GameMode } from "./mode";
import Resource, { ResourceContainer } from "./resource";
import { Stat } from "./stat";

export class APItem extends Resource {}

export const APITEMLIST = [
    new APItem(0, 'energyPotionA', 'Energy Potion α', GameMode.ALL, 0, [[Stat.ENERGY_POWER, 100]]),
    new APItem(1, 'energyPotionB', 'Energy Potion β', GameMode.ALL, 0, [[Stat.ENERGY_POWER, 100]]),
    new APItem(2, 'energyPotionD', 'Energy Potion δ', GameMode.ALL, 0, [[Stat.ENERGY_POWER, 100]]),
    new APItem(3, 'magicPotionA', 'Magic Potion α', GameMode.ALL, 0, [[Stat.MAGIC_POWER, 100]]),
    new APItem(4, 'magicPotionB', 'Magic Potion β', GameMode.ALL, 0, [[Stat.MAGIC_POWER, 100]]),
    new APItem(5, 'magicPotionD', 'Magic Potion δ', GameMode.ALL, 0, [[Stat.MAGIC_POWER, 100]]),
    new APItem(6, 'resource3PotionA', 'Resource 3 Potion α', GameMode.ALL, 0, [[Stat.RES3_POWER, 200]]),
    new APItem(7, 'resource3PotionB', 'Resource 3 Potion β', GameMode.ALL, 0, [[Stat.RES3_POWER, 100]]),
    new APItem(8, 'resource3PotionD', 'Resource 3 Potion δ', GameMode.ALL, 0, [[Stat.RES3_POWER, 200]]),
    new APItem(9, 'energyBarBar', 'Energy Bar Bar', GameMode.ALL, 0, [[Stat.ENERGY_BARS, 100]]),
    new APItem(10, 'magicBarBar', 'Magic Bar Bar', GameMode.ALL, 0, [[Stat.MAGIC_BARS, 100]]),
    new APItem(11, 'macGuffinMuffin', 'MacGuffin Muffin', GameMode.ALL, 0, []),
    new APItem(12, 'icarusProudbottomsHomemadeFertilizer', ' Icarus Proudbottoms Homemade Fertilizer', GameMode.ALL, 0, []),
    new APItem(13, 'littleBluePill', 'Little Blue Pill', GameMode.ALL, 0, []), // Doubles your [[PP]] gain from [[ITOPOD]]. Each pill last for one kill.);
    new APItem(14, 'beastButter', 'Beast Butter', GameMode.ALL, 0, []), // Doubles your [[QP]] reward on your next [[Questing|quest]].);
    new APItem(15, 'luckyCharm', 'Lucky Charm', GameMode.ALL, 0, [[Stat.DROP_CHANCE, 100]]),
    new APItem(16, 'superLuckyCharm', 'Super Lucky Charm', GameMode.ALL, 0, [[Stat.DROP_CHANCE, 100]]),
    new APItem(17, 'mayoInfuser', 'Mayo Infuser', GameMode.ALL, 0, [[Stat.MAYO_SPEED, 100]]),
    new APItem(18, 'regularBlackPens', 'Regular Black Pens', GameMode.ALL, 0, []), // Adds 2 tiers to the next card spawned. Cannot be turned off.);
    new APItem(19, 'improvedLootFilter', 'Improved Loot Filter', GameMode.ALL, 0, []),
    new APItem(20, 'extraInventorySpace', 'Extra Inventory Space', GameMode.ALL, 0, []),
    new APItem(21, 'autoMergeandBoostTimers', '1/2 Auto Merge and Boost Timers!', GameMode.ALL, 0, []),
    new APItem(22, 'instaTrainingCap', 'Insta Training Cap', GameMode.ALL, 0, []),
    new APItem(23, 'customEnergy/Magic%Buttons', 'Custom Energy/Magic % Buttons', GameMode.ALL, 0, []),
    new APItem(24, 'moreCustomEnergy/Magic%Buttons', 'More Custom Energy/Magic % Buttons', GameMode.ALL, 0, []),
    new APItem(25, 'yggdrasilHarvestLight', ' Yggdrasil Harvest Light', GameMode.ALL, 0, []),
    new APItem(26, '7DayTimeBankforDailySpin', '7-Day Time Bank for Daily Spin!', GameMode.ALL, 0, []),
    new APItem(27, 'loadoutSlot', 'Loadout Slot!', GameMode.ALL, 0, []),
    new APItem(28, 'anExtraBeardSlot', 'An Extra Beard Slot!', GameMode.ALL, 0, []),
    new APItem(29, 'filterBoostsintoInfinityCube', 'Filter Boosts into Infinity Cube!', GameMode.ALL, 0, []),
    new APItem(30, 'lazyITOPODFloorShifter', 'Lazy ITOPOD Floor Shifter', GameMode.ALL, 0, []),
    new APItem(31, 'extraAccessorySlot', 'Extra Accessory Slot!', GameMode.ALL, 0, []),
    new APItem(32, 'daycareSpeedBoost', 'Daycare Speed Boost!', GameMode.ALL, 0, [[Stat.DAYCARE_TIME, 10]]),
    new APItem(33, 'anotherExtraAccessorySlot', 'Another Extra Accessory Slot!', GameMode.ALL, 0, []),
    new APItem(34, 'diggerSlots', 'Digger Slots!', GameMode.ALL, 0, []),
    new APItem(35, 'aMacGuffinSlot', 'A MacGuffin Slot!', GameMode.ALL, 0, []),
    new APItem(36, 'questReminder', 'Quest Reminder!', GameMode.ALL, 0, []),
    new APItem(37, 'fasterQuesting', 'Faster Questing!', GameMode.ALL, 0, []),
    new APItem(38, 'extendedQuestBank', 'Extended Quest Bank', GameMode.ALL, 0, []),
    new APItem(39, 'anotherExtraAccessorySlot2', 'Another Extra Accessory Slot!', GameMode.ALL, 0, []),
    new APItem(40, 'customIdleEnergyMagicButtons', 'Custom Idle Energy/Magic % Buttons!', GameMode.ALL, 0, []),
    new APItem(41, 'autoNuker', 'Auto Nuker', GameMode.ALL, 0, []),
    new APItem(42, 'yetAnotherExtraAccessorySlot', 'Yet Another Extra Accessory Slot!', GameMode.ALL, 0, []),
    new APItem(43, 'nGUCapModifier', 'NGU Cap Modifier', GameMode.ALL, 0, []),
    new APItem(44, 'daycareKittyArt', 'Daycare Kitty Art!', GameMode.ALL, 0, []),
    new APItem(45, 'customResource3Button', 'Custom Resource 3 % Button!', GameMode.ALL, 0, []),
    new APItem(46, 'anotherCustomResource3Button', 'Another Custom Resource 3 % Button!', GameMode.ALL, 0, []),
    new APItem(47, 'customIdleResource3Button', 'Custom Idle Resource 3 % Button!', GameMode.ALL, 0, []),
    new APItem(48, 'resource3NameRandomizer', 'Resource 3 Name Randomizer', GameMode.ALL, 0, []),
    new APItem(49, 'fasterWishes', 'Faster Wishes', GameMode.ALL, 0, [[Stat.WISH_SPEED, 25]]),
    new APItem(50, 'inventoryMergeSlots', 'Inventory Merge Slots', GameMode.ALL, 0, []),
    new APItem(51, 'adventureLight', 'Adventure Light', GameMode.ALL, 0, []),
    new APItem(52, 'adventureAdvancer', 'Adventure Advancer', GameMode.ALL, 0, []),
    new APItem(53, 'goToQuestZoneButton', 'Go To Quest Zone Button', GameMode.ALL, 0, []),
    new APItem(54, 'anEvilAccessorySlot', ' An Evil Accessory Slot', GameMode.ALL, 0, []),
    new APItem(55, 'extraDeckSize', 'Extra Deck Size!', GameMode.ALL, 0, []),
    new APItem(56, 'mayoGenerator', 'Mayo Generator', GameMode.ALL, 0, [[Stat.MAYO_GENERATOR, 2]]),
    new APItem(57, 'extraTagSlot', 'Extra Tag Slot!', GameMode.ALL, 0, []),
    new APItem(58, 'extraAccessorySlot2', 'Extra Accessory Slot', GameMode.ALL, 0, []),
    new APItem(59, 'redHeart', 'My Red Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(60, 'yellowHeart', 'My Yellow Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(61, 'brownHeart', 'My Brown Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(62, 'greenHeart', 'My Green Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(63, 'blueHeart', 'My Blue Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(64, 'purpleHeart', 'My Purple Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(65, 'orangeHeart', 'My Orange Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(66, 'greyHeart', 'My Grey Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(67, 'pinkHeart', 'My Pink Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(68, 'rainbowHeart', 'My Rainbow Heart &lt;3', GameMode.ALL, 0, []),
    new APItem(69, '200EXP', '200 EXP!', GameMode.ALL, 0, []),
    new APItem(70, '500EXP', '500 EXP!', GameMode.ALL, 0, []),
    new APItem(71, '2000EXP', '2000 EXP!', GameMode.ALL, 0, []),
    new APItem(72, '25PP', ' 25 PP!', GameMode.ALL, 0, []),
    new APItem(73, '100PP', '100 PP!', GameMode.ALL, 0, []),
    new APItem(74, '500PP', '500 PP!', GameMode.ALL, 0, []),
];

export var APITEMS = new ResourceContainer(APITEMLIST);