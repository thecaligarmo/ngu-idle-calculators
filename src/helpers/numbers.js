export function bigdec_max(...args){
    if (args.length < 1){ 
        throw 'Max of empty list';
    }
    var m = args[0];
    args.forEach( a => {
        if (m.compareTo(a) < 0) {
            m = a
        }
    });
    return m;
}

export function bigdec_min(...args){
    if (args.length < 1){ 
        throw 'Max of empty list';
    }
    var m = args[0];
    args.forEach( a => {
        if (m.compareTo(a) > 0) {
            m = a
        }
    });
    return m;
}

// print/pretty number
export function pn(num) {
    return num.getPrettyValue()
}