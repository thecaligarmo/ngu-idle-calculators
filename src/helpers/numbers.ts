import bigDecimal from "js-big-decimal";
import { getLargeSuffix } from "./largeNumbers";
import _ from "lodash";

export function bigdec_max(...args : bigDecimal[]) : bigDecimal {
    if (args.length < 1){ 
        throw 'Max of empty list';
    }
    var m = args[0];
    args.forEach( a => {
        if (m.compareTo(a) < 0) {
            m = a
        }
    });
    return m;
}

export function bigdec_min(...args : bigDecimal[]) : bigDecimal {
    if (args.length < 1){ 
        throw 'Min of empty list';
    }
    var m = args[0];
    args.forEach( a => {
        if (m.compareTo(a) > 0) {
            m = a
        }
    });
    return m;
}

// print/pretty number
export function pn(num : bigDecimal, numberFormat : string = 'scientific', precision : number = 0) : string {
    // Get the floor of the number for processing
    var n = num.floor().getValue();
    var nl = n.length // number length

    // We should only start for billions as until then it's pretty human readable
    if (nl < 10) {
        if (nl < 3 && precision == 0) {
            precision = 2
        }
        return Number(num.round(precision).getValue()).toLocaleString() //.toFixed(precision).toLocaleString()
    }
    
    // If we're in scientific format, we only want the first 4 digits
    if (numberFormat == 'scientific') {
        var e = nl - 1
        var firstFour = n.slice(0,4);
        return (Number(firstFour) / 1000).toFixed(3).toLocaleString() + "E+" + e
    }

    // Number of digits in front
    var nmod = nl % 3
    // Number of 3 numbers
    var e = (nmod == 0) ? Math.floor(nl / 3) - 1 : Math.floor(nl / 3);
    var firstNum = (nmod == 0) ? n.slice(0, 6) : n.slice(0, nmod + 3);
    if (numberFormat == 'engineering') {
        return (Number(firstNum) / 1000).toFixed(3).toLocaleString() + "E+" + (e * 3)
    }
    var suffix = getLargeSuffix(e-1)
    return (Number(firstNum) / 1000).toFixed(3).toLocaleString() + " " + suffix
}


export function bd(num : any) : bigDecimal {
    if(num instanceof bigDecimal) {
        return num
    }
    return new bigDecimal(num)
}

// date number
export function dn(num : bigDecimal) : string{
    if ( _.isUndefined(num) || num.compareTo(bd(0)) == -1) {
        return '00:00'
    }
    var day = bd(3600 * 24)
    var hour = bd(3600)
    var minute = bd(60)
    var num = num.floor()
    var d = num.divide(day).floor();
    var h = num.modulus(day).divide(hour).floor();
    var m = num.modulus(hour).divide(minute).floor();
    var s = num.modulus(minute).floor();

    var dString = d.getValue() + ":"
    var hString = h.getValue().padStart(2, '0') + ":"
    var mString = m.getValue().padStart(2, '0') + ":"
    var sString = s.getValue().padStart(2, '0')

    var str = mString + sString
    if (h.compareTo(bd(0)) == 1 || d.compareTo(bd(0)) == 1) {
        str = hString + str
    }
    if (d.compareTo(bd(0)) == 1) {
        str = dString + str
        if(d.compareTo(bd(365)) == 1) {
            return 'âˆž'
        }
    }
    return str
}


