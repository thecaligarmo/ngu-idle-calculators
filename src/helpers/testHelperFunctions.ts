
import { bd, toNum } from '@/helpers/numbers'
import bigDecimal from 'js-big-decimal'

export interface infoObjType {
    'ap' ?: number,
    'attack' ?: number,
    'bar' ?: number,
    'blood' ?: number,
    'boosts' ?: number,
    'cap' ?: number,
    'cardSpeed' ?: number,
    'daycare' ?: number,
    'dc' ?: number,
    'energy' ?: number,
    'ecap' ?: number,
    'engu' ?: number,
    'epower' ?: number,
    'exp' ?: number,
    'gold' ?: number,
    'hack' ?: number,
    'magic' ?: number,
    'mayoGen' ?: number,
    'mayoSpeed' ?: number,
    'mcap' ?: number,
    'mngu' ?: number,
    'mwandoos' ?: number,
    'power' ?: number,
    'pp' ?: number,
    'qp' ?: number,
    'respawn' ?: number,
    'reward' ?: number,
    'rpower' ?: number,
    'tagEffect' ?: number,
    'toughness' ?: number,
    'wish' ?: number,
    'ygg' ?: number,
}

export function expectClose(value : number | bigDecimal , expect : number | bigDecimal, range ?: number) : number[]{
    value = toNum(value)
    expect = toNum(expect)
    
    if (range == null) {
        range = Math.max(bd(expect).floor().getValue().length - 3, 0)
    }

    const v = value / (10 ** range)
    const e = expect / (10 ** range)
    return [v, e]
}


// export function toDataObj(playerData : any, extraRequired : any[] = []) : {[key:string] : any}{
//     var playerNumberOptions = getPlayerNumberOptions();
//     var playerOptions = getPlayerOptions();
//     var calculatedOptions  = getCalculatedOptions();
//     for (var col of extraRequired) {
//         for (var k of col) {
//             playerNumberOptions.push(k)
//         }
//     }
//     var dataObj : {[key :string] : any}= {}
//     for (var key of playerNumberOptions) {
//         var defaultVal = defaultPlayerData(playerData, key)
//         dataObj[key] = [defaultVal]
//     }
//     for (var key of playerOptions) {
//         var defaultVal = defaultPlayerData(playerData, key)
//         dataObj[key] = [defaultVal]
//     }

//     for (var key of calculatedOptions) {
//         var defaultVal = defaultPlayerData(dataObj, key).getValue()
//         dataObj[key] = [defaultVal]
//     }

//     return dataObj;
// }