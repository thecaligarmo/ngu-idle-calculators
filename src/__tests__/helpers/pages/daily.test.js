import earlyEvilTwo from '@/__tests__/__data__/earlyEvil2';
import earlyNormal from '@/__tests__/__data__/earlyNormal1';
import earlyNormalTwo from '@/__tests__/__data__/earlyNormal2';
import lateNormal from '@/__tests__/__data__/lateNormal';
import midEvil from '@/__tests__/__data__/midEvil1';
import midNormal from '@/__tests__/__data__/midNormal1';
import midNormalTwo from '@/__tests__/__data__/midNormal2';
import { expectClose, toDataObj } from '@/__tests__/testHelperFunctions';
import { AttackStat, Titans } from "@/assets/enemy";
import { bd } from '@/helpers/numbers';
import { getDailySaveAP, getDailySpinAP, getMaxTitanByAK, getMoneyPitAP, getQuestInfo, getRebirthAP, getTitanHourlyInfo, getTitanList } from '@/helpers/pages/daily';
import renderer from 'react-test-renderer';
import earlyEvil from '@/__tests__/__data__/earlyEvil1';
import midEvilTwo from '@/__tests__/__data__/midEvil2';
import lateEvil from '@/__tests__/__data__/lateEvil1';
import earlySad from '@/__tests__/__data__/earlySad1';



var earlyNormalData = toDataObj(earlyNormal);
var earlyNormalTwoData = toDataObj(earlyNormalTwo);
var midNormalData = toDataObj(midNormal);
var midNormalTwoData = toDataObj(midNormalTwo);
var lateNormalData = toDataObj(lateNormal);
var lateEvilData = toDataObj(lateEvil);
var earlyEvilData = toDataObj(earlyEvil);
var earlyEvilTwoData = toDataObj(earlyEvilTwo);
var midEvilData = toDataObj(midEvil);
var midEvilTwoData = toDataObj(midEvilTwo);
var earlySadData = toDataObj(earlySad);

// getQuestInfo

describe("Daily page - Quest Info", () => {
    var cases = [
        [lateNormalData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : false,
            includeMajorQuests : false,
        }, {
            'ap' : {'major' : bd(0), 'minor': bd(75), 'perMajor' : bd(173), 'perMinor' : bd(17)},
            'qp' : {'major' : bd(0), 'minor': bd(48), 'perMajor' : bd(114), 'perMinor' : bd(11)},
        }],
        [earlyEvilData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : false,
            includeMajorQuests : true,
        }, {
            'ap' : {'major' : bd(708), 'minor': bd(107), 'perMajor' : bd(185), 'perMinor' : bd(18)},
            'qp' : {'major' : bd(536), 'minor': bd(83), 'perMajor' : bd(140), 'perMinor' : bd(14)},
        }],
        [earlyEvilTwoData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : true,
            includeMajorQuests : true,
        }, {
            'ap' : {'major' : bd(356), 'minor': bd(108), 'perMajor' : bd(93), 'perMinor' : bd(18)},
            'qp' : {'major' : bd(306), 'minor': bd(96), 'perMajor' : bd(80), 'perMinor' : bd(16)},
        }],
        [midEvilTwoData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : false,
            includeMajorQuests : false,
        }, {
            'ap' : {'major' : bd(0), 'minor': bd(239), 'perMajor' : bd(191), 'perMinor' : bd(22)},
            'qp' : {'major' : bd(0), 'minor': bd(554), 'perMajor' : bd(428), 'perMinor' : bd(51)},
        }],
        [lateEvilData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : false,
            includeMajorQuests : false,
        }, {
            'ap' : {'major' : bd(0), 'minor': bd(385), 'perMajor' : bd(229), 'perMinor' : bd(22)},
            'qp' : {'major' : bd(0), 'minor': bd(1265), 'perMajor' : bd(725), 'perMinor' : bd(72)},
        }],
        [earlySadData, {
            beastButter : false,
            hoursOfflinePerDay : bd(0),
            hoursPerDay : bd(24),
            idleMajorQuests : false,
            includeMajorQuests : false,
        }, {
            'ap' : {'major' : bd(0), 'minor': bd(412), 'perMajor' : bd(230), 'perMinor' : bd(23)},
            'qp' : {'major' : bd(0), 'minor': bd(2440), 'perMajor' : bd(1368), 'perMinor' : bd(136)},
        }],
    ]
    test.each(cases)(
        "Daily Page - Quest Info - Case %#",
        (data, extraData, expectedValue) => {

            var infoData = {
                totalAPBonus: bd(data['totalAPBonus%'][0]),
                totalQPBonus: bd(data['totalQuestRewardBonus%'][0]),
                activeQuestWishI: bd(data['activeQuestWishI-2'][0]),
                activeQuestWishII: bd(data['activeQuestWishII-2'][0]),
                blueHeart: data['blueHeart^'][0] == 1,
                fadLandsSetBonus: data['fadLandsSetBonus^'][0] == 1,
                fasterQuesting: data['fasterQuesting^'][0] == 1,
                fibQuestRNG: data['fibQuestRNG^'][0] == 1,
                redLiquidBonus: data['redLiquidBonus^'][0] == 1,
                questIdleDivider: bd(data['questIdleDivider-1'][0]),
                questMinorQP: bd(data['questMinorQP-2'][0]),
                questMajorQP: bd(data['questMajorQP-2'][0]),
                totalQuestDropBonus: bd(data['totalQuestDropBonus%'][0]),
                totalRespawnTime: bd(data['totalRespawnTime'][0]),
            }


            var combinedData = {...infoData, ...extraData}
            var questInfo = getQuestInfo(combinedData)
            for (var t of ['ap', 'qp']){
                for (var m of ['major', 'minor', 'perMajor', 'perMinor']) {
                    var ec = expectClose(questInfo[t][m], expectedValue[t][m])
                    expect(ec[0]).toBeCloseTo(ec[1], 0)
                }
            }
        }
    )
})



test('Daily Page - Title List', ()=> {
    var titanList = getTitanList()
    const titanListSnap = renderer.create(<>{titanList}</>).toJSON();
    expect(titanListSnap).toMatchSnapshot();
})

describe("Daily page - Max Titan AK", () => {
    var cases = [
        [earlyNormalData, [Titans.NONE, 0]], // 0
        [earlyNormalTwoData, [Titans.GRAND_TREE, 0]],
        [midNormalData, [Titans.UUG, 0]], // 2
        [midNormalTwoData, [Titans.WALDERP, 4]],
        [lateNormalData, [Titans.BEAST, 2]], // 4
        [earlyEvilData, [Titans.BEAST, 3]],
        [earlyEvilTwoData, [Titans.BEAST, 3]], // 6
        [midEvilData, [Titans.NERD, 1]],
        [midEvilTwoData, [Titans.GODMOTHER, 1]], // 8
        [lateEvilData, [Titans.EXILE, 0]],
        [earlySadData, [Titans.EXILE, 3]],
    ]
    test.each(cases)(
        "Daily Page - Max Titan AK - Case %#",
        (data, expectedValue) => {
            var playerAttack = new AttackStat(1, bd(data['totalPower'][0]), bd(data['totalToughness'][0]), bd(data['totalRegen'][0]), bd(data['totalHealth'][0]))
            var titans = data['titans'][0]
            var maxTitanByAK = getMaxTitanByAK(titans, playerAttack)
            expect(maxTitanByAK[0].key).toBe(expectedValue[0].key)
            expect(maxTitanByAK[1]).toBe(expectedValue[1])
        }
    )
})

describe("Daily page - Titan Hourly Info", () => {
    var cases = [
        [[Titans.NONE, 0], earlyNormalData, {
            'ap' : bd(0),
            'exp' : bd(0),
            'ppp' : bd(0),
            'qp' : bd(0),
        }],
        [[Titans.GRAND_TREE, 0], earlyNormalTwoData, {
            'ap' : bd(25),
            'exp' : bd(95),
            'ppp' : bd(0),
            'qp' : bd(0),
        }],
        [[Titans.UUG, 0], midNormalData, {
            'ap' : bd(100.57),
            'exp' : bd(685),
            'ppp' : bd(0),
            'qp' : bd(0),
        }],
        [[Titans.WALDERP, 4], midNormalTwoData, {
            'ap' : bd(262),
            'exp' : bd(2172),
            'ppp' : bd(0),
            'qp' : bd(0),
        }],
        [[Titans.BEAST, 2], lateNormalData, {
            'ap' : bd(354),
            'exp' : bd(78657),
            'ppp' : bd(871450),
            'qp' : bd(0),
        }],
        [[Titans.BEAST, 3], earlyEvilData, {
            'ap' : bd(377),
            'exp' : bd(82194),
            'ppp' : bd(3714700),
            'qp' : bd(0),
        }],
        [[Titans.BEAST, 3], earlyEvilTwoData, {
            'ap' : bd(382),
            'exp' : bd(135653),
            'ppp' : bd(6257075),
            'qp' : bd(0),
        }],
        [[Titans.GODMOTHER, 1], midEvilTwoData, { // Per hour, off by one for version
            'ap' : bd(9360/24), 
            'exp' : bd(77909040/24),
            'ppp' : bd(9152102400/24),
            'qp' : bd(120/24),
        }],
        [[Titans.EXILE, 0], lateEvilData, { // Per hour, off by one for version
            'ap' : bd(9360/24), 
            'exp' : bd(74626956/24),
            'ppp' : bd(32276551680/24),
            'qp' : bd(696/24),
        }],
        
        [[Titans.EXILE, 3], earlySadData, { // Per hour, off by one for version
            'ap' : bd(9408/24), 
            'exp' : bd(528321348/24),
            'ppp' : bd(102675384240/24),
            'qp' : bd(2400/24),
        }],
    ]
    test.each(cases)(
        "Daily Page - Titan Hourly Info - Case %#",
        (maxTitan, data, expectedValue) => {

            var infoData = {
                totalAPBonus: bd(data['totalAPBonus%'][0]),
                totalExpBonus: bd(data['totalExpBonus%'][0]),
                totalPPBonus : bd(data['totalPPBonus%'][0]),
                totalQPBonus: bd(data['totalQuestRewardBonus%'][0]),
                bonusTitanEXPPerk: bd(data['bonusTitanEXPPerk-2'][0]),
                numRebirthChallenges: bd(data['numRebirthChallenges-2'][0]),
                twentyFourHourChallenge: bd(data['twentyFourHourChallenge-2'][0]),
                twentyFourHourEvilChallenge: bd(data['twentyFourHourEvilChallenge-2'][0]),
                twentyFourHourSadisticChallenge: bd(data['twentyFourHourSadisticChallenge-2'][0]),
                wishTitansHadBetterRewards : bd(data['wishTitansHadBetterRewards-2'][0]),
                wishes : data['wishes'][0],
            }

            var titanHourlyInfo = getTitanHourlyInfo(maxTitan, infoData)
            for (var t of ['ap', 'exp', 'ppp', 'qp']){
                var ec = expectClose(titanHourlyInfo[t], expectedValue[t])
                expect(ec[0]).toBeCloseTo(ec[1], 0)
            }
        }
    )
})


describe("Daily page - AP Rebirth", () => {
    var cases = [
        [bd(124.7), bd(24), 206],
        [bd(158.5), bd(20), 216],
        [bd(175.35), bd(15), 176],
    ]
    test.each(cases)(
        "Daily Page - AP Rebirth - Case %#",
        (totalAPBonus, hoursPerDay, expectedValue) => {
            var apRebirth = getRebirthAP(totalAPBonus, hoursPerDay,)
            expect(Number(apRebirth.getValue())).toBeCloseTo(expectedValue)
        }
    )
})


describe("Daily page - AP Money Pit", () => {
    var cases = [
        [bd(50), bd(4), bd(175.35), 348],
        [bd(22), bd(6), bd(125.7), 162],
        [bd(25), bd(2), bd(150.15), 74],
    ]
    test.each(cases)(
        "Daily Page - AP Money Pit - Case %#",
        (goldToss, numTosses, totalAPBonus, expectedValue) => {
            var apMoneyPit = getMoneyPitAP(goldToss, numTosses, totalAPBonus)
            expect(Number(apMoneyPit.getValue())).toBeCloseTo(expectedValue)
        }
    )
})


describe("Daily page - AP Daily Save", () => {
    var cases = [
        [bd(124.7), bd(24), 249],
        [bd(158.5), bd(20), 264.1667],
        [bd(175.35), bd(15), 218.75],
    ]
    test.each(cases)(
        "Daily Page - AP Daily Save - Case %#",
        (totalAPBonus, hoursPerDay, expectedValue) => {

            var apDailySave = getDailySaveAP(totalAPBonus, hoursPerDay)
            expect(Number(apDailySave.getValue())).toBeCloseTo(expectedValue)
        }
    )
})


describe("Daily page - AP Daily Spin", () => {
    var cases = [
        [1, bd(102), bd(24), false, false, 227],
        [2, bd(124.7), bd(24), true, true, 1080],
        [3, bd(124.7), bd(24), true, true, 2060],
        [4, bd(102), bd(24), false, false, 1020],
        [4, bd(102), bd(24), false, true, 1402],
        [4, bd(124.7), bd(24), true, true, 4798],
        [5, bd(124.7), bd(20), true, false, 3128],
        [5, bd(124.7), bd(24), true, true, 6489],
        [6, bd(159.65), bd(22), false, true, 4344],
        [6, bd(124.7), bd(24), true, true, 11290],
        [7, bd(159.65), bd(22), false, true, 5340],
        [7, bd(124.7), bd(24), true, true, 13851],
    ]
    test.each(cases)(
        "Daily Page - AP Daily Spin - Case %#",
        (dailySpinTier, totalAPBonus, hoursPerDay, incConsumables, incJackpot, expectedValue) => {

            var apDailySpin = getDailySpinAP(dailySpinTier, totalAPBonus, hoursPerDay, incConsumables, incJackpot)
            expect(Number(apDailySpin.getValue())).toBeCloseTo(expectedValue, 0)
        }
    )
})