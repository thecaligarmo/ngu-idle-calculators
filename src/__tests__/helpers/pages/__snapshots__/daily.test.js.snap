// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Daily Page - Title List 1`] = `
[
  <option
    value="0-0"
  >
    No Titan
  </option>,
  <option
    value="1-0"
  >
    Gordon Ramsay Bolton
  </option>,
  <option
    value="2-0"
  >
    Grand Corrupted Tree
  </option>,
  <option
    value="3-0"
  >
    Jake From Accounting
  </option>,
  <option
    value="4-0"
  >
    UGG, The Unmentionable
  </option>,
  <option
    value="5-0"
  >
    Walderp
  </option>,
  <option
    value="6-0"
  >
    The Beast v1
  </option>,
  <option
    value="6-1"
  >
    The Beast v2
  </option>,
  <option
    value="6-2"
  >
    The Beast v3
  </option>,
  <option
    value="6-3"
  >
    The Beast v4
  </option>,
  <option
    value="7-0"
  >
    Greasy Nerd v1
  </option>,
  <option
    value="7-1"
  >
    Greasy Nerd v2
  </option>,
  <option
    value="7-2"
  >
    Greasy Nerd v3
  </option>,
  <option
    value="7-3"
  >
    Greasy Nerd v4
  </option>,
  <option
    value="8-0"
  >
    The Godmother v1
  </option>,
  <option
    value="8-1"
  >
    The Godmother v2
  </option>,
  <option
    value="8-2"
  >
    The Godmother v3
  </option>,
  <option
    value="8-3"
  >
    The Godmother v4
  </option>,
  <option
    value="9-0"
  >
    The Exile v1
  </option>,
  <option
    value="9-1"
  >
    The Exile v2
  </option>,
  <option
    value="9-2"
  >
    The Exile v3
  </option>,
  <option
    value="9-3"
  >
    The Exile v4
  </option>,
  <option
    value="10-0"
  >
    IT HUNGERS v1
  </option>,
  <option
    value="10-1"
  >
    IT HUNGERS v2
  </option>,
  <option
    value="10-2"
  >
    IT HUNGERS v3
  </option>,
  <option
    value="10-3"
  >
    IT HUNGERS v4
  </option>,
  <option
    value="11-0"
  >
    Rock Lobster v1
  </option>,
  <option
    value="11-1"
  >
    Rock Lobster v2
  </option>,
  <option
    value="11-2"
  >
    Rock Lobster v3
  </option>,
  <option
    value="11-3"
  >
    Rock Lobster v4
  </option>,
  <option
    value="12-0"
  >
    Amalgamate v1
  </option>,
  <option
    value="12-1"
  >
    Amalgamate v2
  </option>,
  <option
    value="12-2"
  >
    Amalgamate v3
  </option>,
  <option
    value="12-3"
  >
    Amalgamate v4
  </option>,
]
`;
